psydac/api/discretization.py

```diff
@@ -280,132 +283,143 @@ def solve(self, **kwargs):
 
 #==============================================================================
 class DiscreteDerham(BasicDiscrete):
-    def __init__(self, *spaces):
+    """
+    Rerpresent the discrete De Rham sequence
     
-        dim       = len(spaces) - 1
-        self._V0  = None
-        self._V1  = None
-        self._V2  = None
-        self._V3  = None
-        self._dim = dim
-        
-        if dim == 1:
-            self._V0 = spaces[0]
-            self._V1 = spaces[1]
-        elif dim == 2:
-            self._V0 = spaces[0]
-            self._V1 = spaces[1]
-            self._V2 = spaces[2]
-        elif dim == 3:
-            self._V0 = spaces[0]
-            self._V1 = spaces[1]
-            self._V2 = spaces[2]
-            self._V3 = spaces[3]
-
-    # ...
+    """
+    def __init__(self, *spaces):
+
+        dim          = len(spaces) - 1
+        self._dim    = dim
+        self._spaces = spaces
+
+        if dim not in [1,2,3]:
+            raise ValueError('dimension {} is not available'.format(dim))
+
     @property
     def dim(self):
         return self._dim
 
-    # ...  
     @property
     def V0(self):
-        return self._V0
+        return self._spaces[0]
 
     @property
     def V1(self):
-        return self._V1        
+        return self._spaces[1]
 
     @property
     def V2(self):
-        return self._V2
-        
+        return self._spaces[2]
+
     @property
     def V3(self):
-        return self._V3  
+        return self._spaces[3]
+
+    @property
+    def spaces(self):
+        return self._spaces
+
+    @property
+    def derivatives_as_matrices(self):
+        return tuple(V.diff.matrix for V in self.spaces[:-1])
+
+    @property
+    def derivatives_as_operators(self):
+        return tuple(V.diff for V in self.spaces[:-1])
+
+    def projectors(self, *, kind='global', nquads=None):
+
+        if not (kind == 'global'):
+            raise NotImplementedError('only global projectors are available')
+
+        if self.dim == 1:
+            P0 = Projector_H1(self.V0)
+            P1 = Projector_L2(self.V1, nquads)
+            return P0, P1
+
+        elif self.dim == 2:
+            raise NotImplementedError('TODO')
+
```

Comment:
It should not be a problem to provide the projectors in 2D, right?

Reply:
yes the only problem is we need to decide which sequence the user wants in 2d

Reply:
Sure, but the user tells us explicitly which 2D DeRham sequence is needed when the symbolic DeRham sequence is created.

Reply:
yes that's true, we can add it